### DISCUSSION TITLE
What is a keystore?

### DISCUSSION THREAD

**Discussion by talasjanos** [keycloak/keycloak#14622] [Q&A] [answered]
Hi Team,

As [this guide](https://www.keycloak.org/server/containers) explains, the only way of providing the SSL certificates to Keycloak is to give them as a keystore file (with a password). I searched but could not find any other guide referencing what this keystore would be and how to generate it from existing HTTPS certificate files. Could you please refer me to such guide?

Or alternatively, is there any (undocumented) way of providing those certificates directly?

**Comment by darius-m** [comment, marked as answer]
A keystore / truststore is simply a store for cryptographic keys usually used in Java environments (like the Keycloak backend). You can create a keystore using the `keytool` command line utility that usually comes with the java packages. You can read the manual look up some usage examples for the `keytool` utility to get started (I usually interact really infrequently with it, so I do not have any quick examples).

The difference between keystores and truststores is what they store - the keystore contains private keys and certificates usually used by a server, while the truststore contains public certificates that a client trusts (the server / client roles are not exclusive).

**Reply by talasjanos** [reply]
Thank you for the explanation! Your prompt response is much appreciated.

Meanwhile I realised that the entry point of the docker container is the regular `kc.sh` start script so I can provide the `--https-certificate-file` / `--https-certificate-key-file` command line argument just the same (as described [here](https://www.keycloak.org/server/enabletls)).

**Reply by talasjanos** [reply]
Actually, I cannot. The container is defined to run as user 1000 who has no permission to the certificate files.

**Reply by darius-m** [reply]
The user running Keycloak will need to have access to the certificate files regardless of format. Even if you add them to the keystore, Keycloak will need to be able to access that file; there is no getting around it.

**Reply by talasjanos** [reply]
Well…

```shell
docker run --detach \
  --env KEYCLOAK_ADMIN=admin \
  --env KEYCLOAK_ADMIN_PASSWORD=*REDUCTED* \
   -v /etc/letsencrypt/:/etc/letsencrypt/:ro \
   -v /var/log/keycloak/:/var/log/keycloak/:rw \
   -p 8443:8443 \
  --user root \
  quay.io/keycloak/keycloak:19.0.2 \
  start \
  --https-certificate-file=/etc/letsencrypt/live/*REDUCTED*/fullchain.pem \
  --https-certificate-key-file=/etc/letsencrypt/live/*REDUCTED*/privkey.pem \
  --log-level=WARN \
  --log=console,file \
  --log-file=/var/log/keycloak/keycloak.log \
  --db=mysql \
  --db-url=jdbc:mysql://*REDUCTED*:*REDUCTED*/keycloak \
  --db-username=keycloak \
  --db-password=*REDUCTED* \
  --hostname=*REDUCTED* \
```

The `--user root` did the trick. NGInx (running as user nginx within its default container image) did not have this problem, it did have access to the certificates. Do you happen to know if Keycloak intentionally has such limitation?

**Reply by darius-m** [reply]
The `--user root` parameter starts Keycloak with root permissions inside the container. While the root user inside a container should be different from the root on the host, it may still be a good idea to not run applications as the container root. Since containers are basically just various types of namespaces (network, process, etc.) inside the same OS as the host, the root user, who receives some elevated privileges through capabilities, may in some cases be able to access or alter resources it should not (it is, of course, not trivial).

In your case, I assume that there may be an issue with the permissions on the mapped volumes that keeps the user with ID 1000 from accessing the certificate files. I believe `certbot` creates the certificate files without read permissions for `others` on the host, and the same permissions are kept when mapping the directories inside the container, which means that the user with ID 1000 cannot access them.

The reason nginx may not have the same issue is because (at least when installed as a bare-metal service) it appears to start as the root user, opens the certificate and log files, opens the sockets (e.g. 80 and 443 which are privileged ports), and then forks and changes the user to drop privileges. You should be able to confirm this when inspecting the processes that run on the system - there should be one nginx process marked as `master` that runs as root, and a few other processes that run as a different user (e.g. nginx). Keycloak does not fork to drop privileges, so it starts as the specified user directly, which means that it cannot access files that are not explicitly allowed for the user.

As a somewhat easy workaround, if you are using a reverse proxy (e.g. nginx), you do not need to use the "proper" certificates for Keycloak. Since the connections from clients are (usually) terminated on the reverse proxy, you can use the certificates on the proxy, and encrypt the traffic between the proxy and the Keycloak backends using self-signed certificates. In this case, you would need to configure Keycloak to use a self-signed certificate + key, and have the reverse proxy trust the certificate (copy Keycloak's public certificate somewhere that can be accessed by the proxy and add the proxy-specific directive to trust it - for nginx you would need to add `proxy_ssl_trusted_certificate /path/to/public/cert.pem` in the appropriate `location` block).
